### ISSUE TITLE
Base Map missing Tiling & Offset controls

### ISSUE THREAD

**Issue by vegmuffin** [Unity-Technologies/com.unity.toonshader#298]
**Describe the bug**
The base diffuse map for the toon material is missing tiling & offset parameters, while Normal, MatCap and Emission maps have this control.

**To Reproduce**
1. Install toonshader package
2. Create a new material based on the `Toon` shader
3. Scroll down to `Three Color and Control Map Settings`

**Expected behavior**
The Base Map and other shading maps should have tiling & offset controls.

**Desktop (please complete the following information):**
 - OS: Windows

**Additional context**
I don't know if this is how it should be, but it definitely is not ideal to not have tiling & offset controls, I want to be able to tile textures. :(

**Comment by H3idi-X** [comment]
Thanks for your feedback!
We will adopt features of Unity features, not just UTS, depending on the severity and number of requests.
We currently receive few requests for this feature, but if we receive a large number of requests, we will consider implementing this. Therefore, we will leave this issue here.

**Comment by dustinpb97** [comment]
Hi, I'd also like the ability to tile textures on the toon shader.

**Comment by Fabrazz** [comment]
+1 Would be nice for those values to be public.

You can do this via Debug Inspector and we have been doing it that way for the entirety of our current project.

**Comment by Fabrazz** [comment]
I think it looks perfectly fine!

**Comment by H3idi-X** [comment]
Understood. It would not be ideal to have you use Debug Inspector to deal with the situation. We can't use much time right now, but I will consider the solution.

**Comment by H3idi-X** [comment]
@Hum9183, I started implementing your fix, which is essentially correct. However, we need to apply the property to both the 1st and 2nd shading maps. I believe using one sampler for the Base, 1st shading map, and 2nd shading map textures would be a good approach. Implementing this across all the render pipeline shaders will take some time.

**Comment by H3idi-X** [comment]
Some are done, but some look not working. we are on investigation now.

> one sampler for the Base, 1st shading map, and 2nd shading map textures would be a good approach

**Comment by Hum9183** [comment]
@H3idi-X 
That is a good idea!
If Tilling&Offset is to be shared between BaseMap, 1stShading and 2ndShading, it would be quicker to set `_MainTex` as the texture to be passed to `TRANSFORM_TEX` macro as shown below.
``` diff
--  SAMPLE_TEXTURE2D(texture2d, sampler, TRANSFORM_TEX(uv, _1st_ShadeMap))
++  SAMPLE_TEXTURE2D(texture2d, sampler, TRANSFORM_TEX(uv, _MainTex))
```

I think the actual UTS can handle this by doing the following

In UniversalToonBodyDoubleShadeWithFeather.hlsl (line 142)
``` diff
#ifdef _IS_PASS_FWDBASE
    float3 Set_LightColor = lightColor.rgb;
    float3 Set_BaseColor = lerp( (_BaseColor.rgb*_MainTex_var.rgb), ((_BaseColor.rgb*_MainTex_var.rgb)*Set_LightColor), _Is_LightColor_Base );
    //v.2.0.5
--  float4 _1st_ShadeMap_var = lerp(SAMPLE_TEXTURE2D(_1st_ShadeMap,sampler_MainTex, TRANSFORM_TEX(Set_UV0, _1st_ShadeMap)),_MainTex_var,_Use_BaseAs1st);
++  float4 _1st_ShadeMap_var = lerp(SAMPLE_TEXTURE2D(_1st_ShadeMap,sampler_MainTex, TRANSFORM_TEX(Set_UV0, _MainTex)),_MainTex_var,_Use_BaseAs1st);
    float3 Set_1st_ShadeColor = lerp( (_1st_ShadeColor.rgb*_1st_ShadeMap_var.rgb), ((_1st_ShadeColor.rgb*_1st_ShadeMap_var.rgb)*Set_LightColor), _Is_LightColor_1st_Shade );
    //v.2.0.5
--  float4 _2nd_ShadeMap_var = lerp(SAMPLE_TEXTURE2D(_2nd_ShadeMap, sampler_MainTex, TRANSFORM_TEX(Set_UV0, _2nd_ShadeMap)),_1st_ShadeMap_var,_Use_1stAs2nd);
++  float4 _2nd_ShadeMap_var = lerp(SAMPLE_TEXTURE2D(_2nd_ShadeMap, sampler_MainTex, TRANSFORM_TEX(Set_UV0, _MainTex)),_1st_ShadeMap_var,_Use_1stAs2nd);
    float3 Set_2nd_ShadeColor = lerp( (_2nd_ShadeColor.rgb*_2nd_ShadeMap_var.rgb), ((_2nd_ShadeColor.rgb*_2nd_ShadeMap_var.rgb)*Set_LightColor), _Is_LightColor_2nd_Shade );
    float _HalfLambert_var = 0.5*dot(lerp( i.normalDir, normalDirection, _Is_NormalMapToBase ),lightDirection)+0.5;
```

**Comment by H3idi-X** [comment]
Oh, sorry for the confusion. What I wanted to discuss was the lines where we use TRANSFORM_TEX. Thank you for the correction and suggestions.

**Comment by Hum9183** [comment]
@H3idi-X 
Let's see...
Am I right in saying that I want to keep the UVs of the `_MainTex` in a variable and use them in the 1stShading and 2ndShading calculations?
My apologies if I misunderstood.
``` 
float2 mainTexUV = TRANSFORM_TEX(Set_UV0, _MainTex);
float4 _1st_ShadeMap_var = SAMPLE_TEXTURE2D(_1st_ShadeMap, sampler_MainTex, mainTexUV);
float4 _2nd_ShadeMap_var= SAMPLE_TEXTURE2D(_2nd_ShadeMap,  sampler_MainTex, mainTexUV);
```

**Comment by H3idi-X** [comment]
We adopted the formor ones for the fix.
We are testing the shaders with graphics test systems.

**Comment by H3idi-X** [comment]
merged to the main branch.

**Comment by H3idi-X** [comment]
Hi, we released toon shader 0.9.7-preview this morning.
Will you guys please try that one?

**Comment by H3idi-X** [comment]
Thanks for checking! We will take a look at the two issues above.

**Comment by H3idi-X** [comment]
Hi, we have applied fixes for the second of the following issues  to [task/clippingMaskSampler](https://github.com/Unity-Technologies/com.unity.toonshader/tree/task/clippingMaskSampler) branch.
Could  @Hum9183  and @Fabrazz  try the branch if it doesn’t make too much trouble?

Fixing the first is difficult due to the design of the Unity Editor for now.

> 1. If you change the tiling value and undo, it doesn't update the values in the component to the previous ones until you deselect & reselect the material.
> 2. Clipping Masks should also take that tiling into account otherwise they don't work.

**Comment by Hum9183** [comment]
@H3idi-X 
That sounds great!
In addition, it may be a good idea to synchronize the ClippingMask of the outline

In UniversalToonOutlime.hlsl (lines 96-98)
```diff
#elif _IS_OUTLINE_CLIPPING_YES
-    float4 _ClippingMask_var = SAMPLE_TEXTURE2D(_ClippingMask, sampler_MainTex, TRANSFORM_TEX(Set_UV0, _ClippingMask));
+    float4 _ClippingMask_var = SAMPLE_TEXTURE2D(_ClippingMask, sampler_MainTex, TRANSFORM_TEX(Set_UV0, _MainTex));
     float Set_MainTexAlpha = _MainTex_var.a;
```

**Comment by H3idi-X** [comment]
Good point!
We fix that.
Thank you!

**Comment by H3idi-X** [comment]
@Hum9183 We applied your suggestion and did the same to the oehter RP shaders to the branch.

**Comment by Hum9183** [comment]
@H3idi-X 
>1. If you change the tiling value and undo, it doesn't update the values in the component to the previous ones until you deselect & reselect the material.

I have a simple idea to solve this problem.
May I make a PR?

**Comment by H3idi-X** [comment]
Thanks for the suggestion.
As long as it doesn't affect edtior peformance,the PR would be welcome. 
PR to [task/clippingMaskSampler](https://github.com/Unity-Technologies/com.unity.toonshader/tree/task/clippingMaskSampler) branch is ideal.

**Comment by H3idi-X** [comment]
Hi, we measured time cunsumed in FindProperties(props).
It is negligible. 
We maerged PR by @Hum9183 and started our internal tests.

**Comment by H3idi-X** [comment]
All the tests passed and the branch is merged into the main branch.

**Comment by H3idi-X** [comment]
Hi, we released 0.10.0-preview that we applied the undo/redo fix by @Hum9183.

**Comment by Fabrazz** [comment]
Can confirm on my end, works as intended now! Thanks!
